### ISSUE TITLE
Help replicating on low-VRAM GPUs (40GB/24GB) - reward fails to increase

### ISSUE THREAD

**Issue by jaylee2000** [yifan123/flow_grpo#172]
Hello,

I am from an academic lab with limited compute resources and am trying to replicate your work. Thank you for making the code available.

### Goal & Hardware Constraints

My goal is to replicate the training results on our available hardware, which is lower VRAM than the 80GB A100s used in the configs.

* **Maximum Setup:** 8x A100 (40GB)
* **Ideal Setup:** 4x RTX4090 (24GB)

### The Problem

I have modified the code to fit within our VRAM constraints (details below). While the training runs successfully without OOM errors and I've adjusted the configuration to match the original global batch size (1152), **I am not observing any stable increase in the PickScore reward.**

<img width="1478" height="631" alt="Image" src="https://github.com/user-attachments/assets/ae2f5eb6-7b87-4426-a03a-28cc3b7b8314" />
<img width="1478" height="631" alt="Image" src="https://github.com/user-attachments/assets/65c9bf0e-31ff-4a89-9295-a757ee2c8acf" />
<img width="1478" height="631" alt="Image" src="https://github.com/user-attachments/assets/2b962aa5-1e70-4b90-b029-e9a8299f628c" />
<img width="490" height="318" alt="Image" src="https://github.com/user-attachments/assets/1353aaac-489a-46f2-9e01-eb971180196c" />

A colleague mentioned that he also struggled to reproduce the results on a different GPU setup and suggested that training might only be stable on the exact 80GB GPU configuration.

### My Questions

1.  Is it expected that training stability or convergence is highly sensitive to the hardware configuration (e.g., number of GPUs, VRAM, interconnect)?
2.  If it *should* be possible, I have made several changes and there are multiple potential points of failure. Could you provide any guidance on which of my changes below is the most likely suspect for causing the reward to stagnate?

### Summary of Changes Made

To get the model to run, I've implemented the following modifications:

1.  **DeepSpeed Integration:**
    * Enabled **DeepSpeed ZeRO-2**.
    * Modified the code to let DeepSpeed handle **gradient clipping** and the **optimizer step**.

2.  **Memory Saving Techniques:**
    * Enabled **gradient checkpointing** for the transformer blocks.
    * **Precomputed prompt embeddings** (at `fp16`) and loaded them during training.
    * This allowed me to **offload the tokenizers and text encoders**, saving significant VRAM.

3.  **Batch Size & Config Adjustments:**
    * With the changes above, I can run `config.train.batch_size = 4` on 40GB GPUs, and `config.train.batch_size = 2` on 24GB GPUs (though memory is extremely tight).
    * To match the original global batch size of 1152 (from 48 * 24), I increased `config.train.gradient_accumulation_steps` and `config.sample.num_batches_per_epoch` accordingly.

4.  **Other Code Modifications:**
    * **Dataset:** Switched from `pickscore` to the `pickscore_sfw` dataset.
    * **Sampler:** Modified the data sampler (similar to [PR #107 in yifan1s/flow_grpo](https://github.com/yifan123/flow_grpo/pull/107)) to allow for more flexible batch configurations. This is very helpful for running on my hardware, though I could revert this change by sticking to `config.train.batch_size = 3` on 8x A100.
    * **StatTracker:** Modified `PerPromptStatTracker` to use `prompt_ids` as keys (to support offloading the text encoder). I could revert this by keeping one text encoder in memory if this is a likely issue.

Any advice you could offer on where to focus my debugging efforts would be greatly appreciated.

Thank you!

**Comment by yifan123** [comment]
Thank you very much for your interest in our work. The config looks fine. Are you currently training FLUX? You can first test it on SD3 to check if the training runs properly. You can start by setting the kl beta to 0.

**Comment by XiangqianHuang** [comment]
> Thank you very much for your interest in our work. The config looks fine. Are you currently training FLUX? You can first test it on SD3 to check if the training runs properly. You can start by setting the kl beta to 0.

您好我也遇到了这样的问题，用flux-kontext模型训练时奖励曲线起伏着下跌：

<img width="780" height="453" alt="Image" src="https://github.com/user-attachments/assets/75620ae5-af9b-4ba8-bb5a-773c784028cc" />

<img width="795" height="450" alt="Image" src="https://github.com/user-attachments/assets/e37eb3a8-ac51-4982-b610-d20d53076e2e" />

<img width="782" height="456" alt="Image" src="https://github.com/user-attachments/assets/b436f9fa-bd8b-4281-bab8-d3c2bfa366f6" />

并且拿训练完的lora权重去推理时，产生的图像全是蓝色噪声，只有把lora_scale降低后才能隐约看到图像：

<img width="282" height="177" alt="Image" src="https://github.com/user-attachments/assets/5340e1d3-7030-4714-969a-6aca666a0d0a" />

<img width="301" height="177" alt="Image" src="https://github.com/user-attachments/assets/984ca75d-d2fd-4ec5-b2f2-c68e404455cd" />

我也是因为显存紧张所以设置了很多降低显存的参数，我不确定是不是因为这些设置导致了生成图像的问题：
def counting_flux_kontext():
    gpu_number = 1  # 改为1（单卡
    config = compressibility()
    config.dataset = os.path.join(os.getcwd(), "dataset/counting_edit")

    # sd3.5 medium
    config.pretrained.model = "black-forest-labs/FLUX.1-Kontext-dev"
    config.sample.num_steps = 6
    config.sample.eval_num_steps = 28
    config.sample.guidance_scale = 2.5

    config.resolution = 512  
    config.sample.train_batch_size = 1  # ✅ 从4改为1
    config.sample.num_image_per_prompt = 1  # ✅ 从4改为1
    config.sample.num_batches_per_epoch = 12  # 从48降到12
    assert config.sample.num_batches_per_epoch % 2 == 0, "Please set config.sample.num_batches_per_epoch to an even number! This ensures that config.train.gradient_accumulation_steps = config.sample.num_batches_per_epoch / 2, so that gradients are updated twice per epoch."
    config.sample.test_batch_size = 4  # ✅ 改为4 # This bs is a special design, the test set has a total of 2048, to make gpu_num*bs*n as close as possible to 2048, because when the number of samples cannot be divided evenly by the number of cards, multi-card will fill the last batch to ensure each card has the same number of samples, affecting gradient synchronization.

    config.train.batch_size = 1  # ✅ 关键：降到1
    config.train.gradient_accumulation_steps = 12  # ✅ 增加到12
    config.train.num_inner_epochs = 1
    config.train.timestep_fraction = 0.99
    config.train.beta = 0
    config.sample.global_std = True
    config.sample.same_latent = False
    config.train.ema = True  # 可以节省约2GB
    config.sample.noise_level = 0.9
    config.mixed_precision = "bf16"

    # ========== 新增：启用LoRA ==========
    config.use_lora = True
    config.lora_rank = 16  # ✅ 降低rank
    # 可选：如果想从已有的LoRA checkpoint继续训练
    # config.train.lora_path = "path/to/lora/checkpoint"
    # ====================================

    # ✅ 启用梯度检查点
    config.activation_checkpointing = True

    config.save_freq = 60 # epoch
    config.eval_freq = 60  
    config.save_dir = 'logs/counting_edit/flux_kontext'
    config.reward_fn = {
        # "clipscore": 0.4,
        "deqa": 0.6,
        "pickscore": 0.4
    }
    config.per_prompt_stat_tracking = True
    return config

期待您的回复，谢谢。

**Comment by jaylee2000** [comment]
@yifan123 I am currently training SD3.5-M. It's a bit early to tell, but removing the KL penalty seems to yield similar dynamics although the PickScore reward is a bit higher. I'll report whether or not the reward increases after training a bit longer.

<img width="489" height="307" alt="Image" src="https://github.com/user-attachments/assets/83ef3344-44f3-4bd1-a245-5328e62845c6" />
<img width="489" height="307" alt="Image" src="https://github.com/user-attachments/assets/9d76b97d-b2ce-4677-874f-8b7d229e80aa" />

@XiangqianHuang Have you successfully replicated the results on a smaller model (i.e. SD3.5-M)?

**Comment by XiangqianHuang** [comment]
@jaylee2000 您好，这是我在SD3.5-M上训练的结果：

<img width="801" height="459" alt="Image" src="https://github.com/user-attachments/assets/6747e1d7-0ee9-48c5-8f47-11bdf202eccd" />

<img width="791" height="453" alt="Image" src="https://github.com/user-attachments/assets/af02982e-ff01-4a49-82e3-f1e36f41fd69" />

我是在geneval这个奖励函数上训练的，可以看到有上升的趋势，但由于显存紧张我只训练了360步，评估了四次（0，120，240，360），并且我也拿训练好的权重测试了一下指标：

<img width="247" height="245" alt="Image" src="https://github.com/user-attachments/assets/9ffe9ccf-aa65-413b-8ddb-353f9b27e48c" />

我觉得如果训练时间够长的话，指标、奖励函数曲线应该还会上升

**Comment by jaylee2000** [comment]
@yifan123 The reward doesn't increase even if I set KL penalty to 0. 

<img width="1457" height="315" alt="Image" src="https://github.com/user-attachments/assets/08197da0-a034-4701-a0b3-169c82525bac" />
<img width="730" height="318" alt="Image" src="https://github.com/user-attachments/assets/b5c1a792-8a6b-4e1b-923a-10efcb99a3af" />

In an effort to debug this issue, I plotted importance sampling ratios at the first epoch when off-policy (2nd policy update).
My hypothesis is that DeepSpeed may be introducing additional numerical errors that destabilize calculation of importance sampling ratio, and thus the clipping mechanism. Is this possible?

<img width="569" height="408" alt="Image" src="https://github.com/user-attachments/assets/de3bd262-6999-4657-b827-94ad79e2d248" />

At the final denoising step, almost all of the data is getting clipped. Is this in line with original behavior?
It does seem in line with the video in the [GRPO-Guard page](https://jingw193.github.io/GRPO-Guard/).

<img width="578" height="464" alt="Image" src="https://github.com/user-attachments/assets/63bad209-ad1a-4ee8-b26a-8cd0d929d155" />

<img width="725" height="312" alt="Image" src="https://github.com/user-attachments/assets/e9c9d5d4-ec31-46bb-bf61-b31c6cca6b03" />

I'll try running the next experiment with DeepSpeed off.

**Comment by jaylee2000** [comment]
Update: With DeepSpeed off, I can still run the OCR experiment on 24GB GPUs, and reward seems to be increasing consistently. My guess is that DeepSpeed was the culprit.
I will try another run on PickScore with DeepSpeed off, using 40GB GPUs.

<img width="489" height="319" alt="Image" src="https://github.com/user-attachments/assets/2cc0c8f0-59c1-4c48-9822-3363500288d5" />

**Comment by yifan123** [comment]
Thank you very much for your interest in our work. If you reduce the number of GPUs, please keep the hyperparameters unchanged. Lowering the batch size or group size may cause the reward to stop increasing. The OCR task is the simplest one, so you can start with it to find suitable batch size and group size settings for a smaller GPU setup.

**Comment by jaylee2000** [comment]
@yifan123 Thank you for your dedication and continued support to this codebase!

After running PickScore with DeepSpeed off, I can confirm that the results of Flow-GRPO can be replicated well on low-VRAM GPUs. By enabling gradient checkpointing, I was able to run the PickScore task on 24GB GPUs too. I used batch settings akin to pickscore_sd3_4gpu. (https://github.com/yifan123/flow_grpo/issues/162 was run with Deepspeed on - please ignore my experiments in that issue as they are misleading.)

To anyone who comes across this issue later: I highly discourage using Deepspeed, as it seems to destabilize training. 

<img width="726" height="315" alt="Image" src="https://github.com/user-attachments/assets/cf2c8801-5ab4-4494-a4e2-f6f3357a141b" />

<img width="718" height="306" alt="Image" src="https://github.com/user-attachments/assets/9ed6855f-dc55-42f3-968f-0c60f0d5420b" />

**Comment by jaylee2000** [comment]
Due to several requests for the low-VRAM implemenation, I provide a link [here](https://github.com/jaylee2000/pcpo/tree/main/flowgrpo).
Running `bash scripts/single_node/grpo.sh` will do the trick.
